### PR TITLE
Improve `DelimiterCase`

### PR DESCRIPTION
**Pull request by filipw01**
I made some changes that pass more test cases than the current solution. I was heavily inspired by `CamelCase` implementation

Let me know if this is something worth investing more time into or if I'm doing some basic mistake and butchering performance for example

Also, what should be the behavior of `string1` conversion 
* `string1` - this is the current behavior
* `string_1` - another valid behavior
* either one customizable with an option object like `camelCase` - in that case there's still a question of what should be the default (most likely `string1` for backwards compatibility)

Related issues:
https://github.com/sindresorhus/type-fest/issues/223
https://github.com/sindresorhus/type-fest/issues/336

### CODE REVIEW

**Comment by filipw01** [comment]
@voxpelli Could you take a look? At the moment I just want to know if this is something I should dig deeper into

**Comment by voxpelli** [comment]
@filipw01 Its not immediately obvious to me what the improvements are, and my experience with these helpers has been that they easily turn into a game of whackamole – any improvement reveals a couple of regressions.

I do see that you ultimate refer to my wish to unify the two: https://github.com/sindresorhus/type-fest/issues/224#issuecomment-925357006

> To support this use case I think we essentially would have to merge the functionality of the two. I can't remember why I didn't do that initially.

But that was ultimately solved through a new option instead: https://github.com/sindresorhus/type-fest/pull/501

I guess my main feedback would be to split this PR into more commits and try to make it as clear what and why the changes made are made, so that one can as easily as possible follow the reasoning in a review

**Comment by filipw01** [comment]
For the no split on number I decided to align with change-case which seems to be the best at handling different edge cases

File: source/split-words.d.ts [hunk 1]
```diff
@@ -36,22 +39,30 @@ export type SplitWords<
...
 			// Fist char of word
-			? SplitWords<RemainingCharacters, FirstCharacter, FirstCharacter>
-			// Case change: non-numeric to numeric, push word
+			? SplitWords<RemainingCharacters, Options, FirstCharacter, FirstCharacter>
+			// Case change: non-numeric to numeric
 			: [false, true] extends [IsNumeric<LastCharacter>, IsNumeric<FirstCharacter>]
-				? [...SkipEmptyWord<CurrentWord>, ...SplitWords<RemainingCharacters, FirstCharacter, FirstCharacter>]
-				// Case change: numeric to non-numeric, push word
+				// Split on number: push word
+				? Options['splitOnNumber'] extends true
```
Comment (filipw01) [inline source/split-words.d.ts:46]: When there's a numeric/non-numeric case change I added and used a new option `splitOnNumber` which depending on how it is set will either
1. When set to true - Like before split words on number
2. When set to false - concat to current word

File: source/screaming-snake-case.d.ts [hunk 2]
```diff
@@ -21,13 +11,13 @@ This can be useful when, for example, converting a camel-cased object property t
...
 const someVariable: ScreamingSnakeCase<'fooBar'> = 'FOO_BAR';
+const someVariableNoSplitOnNumber: ScreamingSnakeCase<'p2pNetwork', {splitOnNumber: false}> = 'P2P_NETWORK';
+
 ```
 
 @category Change case
 @category Template literal
-*/
-export type ScreamingSnakeCase<Value> = Value extends string
-	? IsScreamingSnakeCase<Value> extends true
```
Comment (filipw01) [inline source/screaming-snake-case.d.ts:30]: I can bring this back, but I believe it shouldn't be necessary, especially that we don't use it for other things like snake case or kebab case

File: test-d/kebab-case.ts [hunk 3]
```diff
@@ -15,3 +18,60 @@ expectType<'foo-bar'>(kebabFromSnake);
...
+expectType<'foo2bar'>(kebabFromNumberInTheMiddleNoSplitOnNumber);
+
+const kebabFromNumberInTheMiddleNoSplitOnNumberEdgeCase: KebabCase<'foo2Bar', {splitOnNumber: false}> = 'foo2-bar';
+expectType<'foo2-bar'>(kebabFromNumberInTheMiddleNoSplitOnNumberEdgeCase);
+
+const kebabFromNumberInTheMiddleNoSplitOnNumberEdgeCase2: KebabCase<'foO2bar', {splitOnNumber: false}> = 'fo-o2bar';
+expectType<'fo-o2bar'>(kebabFromNumberInTheMiddleNoSplitOnNumberEdgeCase2);
+
+const kebabFromNumberInTheMiddleNoSplitOnNumberEdgeCase3: KebabCase<'FOO22Bar', {splitOnNumber: false}> = 'foo22-bar';
+expectType<'foo22-bar'>(kebabFromNumberInTheMiddleNoSplitOnNumberEdgeCase3);
```
Comment (filipw01) [inline test-d/kebab-case.ts:77]: I added the same tests to all case types as well as some edge cases visible here. If you have any doubts about the implementation I can add more tests

File: source/kebab-case.d.ts [hunk 4]
```diff
@@ -12,6 +13,7 @@ import type {KebabCase} from 'type-fest';
 // Simple
 
 const someVariable: KebabCase<'fooBar'> = 'foo-bar';
+const someVariableNoSplitOnNumber: KebabCase<'p2pNetwork', {splitOnNumber: false}> = 'p2p-network';
```
Comment (filipw01) [inline source/kebab-case.d.ts:16]: I added this new option on case changing types

File: source/delimiter-case.d.ts [hunk 5]
```diff
@@ -1,54 +1,20 @@
...
-
-@see SplitIncludingDelimiters
-*/
-type StringArrayToDelimiterCase<Parts extends readonly any[], Start extends boolean, UsedWordSeparators extends string, UsedUpperCaseCharacters extends string, Delimiter extends string> =
-	Parts extends [`${infer FirstPart}`, ...infer RemainingParts]
-		? `${StringPartToDelimiterCase<FirstPart, Start, UsedWordSeparators, UsedUpperCaseCharacters, Delimiter>}${StringArrayToDelimiterCase<RemainingParts, false, UsedWordSeparators, UsedUpperCaseCharacters, Delimiter>}`
-		: Parts extends [string]
-			? string
-			: '';
+type DelimiterCaseFromArray<
```
Comment (filipw01) [inline source/delimiter-case.d.ts:6]: This gets the words split and inserts the delimited at the beginning and after every word

Like `['here', 'We', 'Go']` -> `'#here#We#Go'`

File: source/delimiter-case.d.ts [hunk 6]
```diff
@@ -87,13 +54,9 @@ const rawCliOptions: OddlyCasedProperties<SomeOptions> = {
...
-export type DelimiterCase<Value, Delimiter extends string> = string extends Value ? Value : Value extends string
-	? StringArrayToDelimiterCase<
-	SplitIncludingDelimiters<Value, WordSeparators | UpperCaseCharacters>,
-	true,
-	WordSeparators,
-	UpperCaseCharacters,
-	Delimiter
-	>
+ */
+export type DelimiterCase<Value, Delimiter extends string, Options extends SplitWordsOptions = {splitOnNumber: true}> = Value extends string
```
Comment (filipw01) [inline source/delimiter-case.d.ts:58]: Here we
1. `'hereWeGo'` -> SplitWords -> `['here', 'We', 'Go']`
2. `['here', 'We', 'Go']` -> DelimiterCaseFromArray ->  `'#here#We#Go'`
3. `'#here#We#Go'` -> RemoveFirstLetter ->  `'here#We#Go'`
4. `'here#We#Go'` -> Lowercase -> `'here#we#go'`

**Comment by filipw01** [comment]
@voxpelli At this point where all the commits are made I'd prefer not to do any git wizardry splitting into smaller commits. I left some comments for easier review. I hope this will be enough. If something is unclear let me know. 

I tried to add more edge cases to avoid the whack a mole game

**Comment by voxpelli** [comment]
#975 is also touching some of these files. I will not have time to review either, just noting it

**Comment by filipw01** [comment]
@sindresorhus any chance to moce this forward?

File: source/delimiter-case.d.ts (hunk 6 above)
Comment (sindresorhus) [inline source/delimiter-case.d.ts:58]: `SplitWordsOptions = {splitOnNumber: true}` will not allow us to add more options in the future, as if you specify one, you override the existing defaults.

Example: https://github.com/sindresorhus/type-fest/blob/7dfb307e3ecc0e7717de6880395f1de8715c82c1/source/paths.d.ts

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
You need to resolve the merge conflict.

**Comment by sindresorhus** [comment]
And maybe `splitOnNumber` => `splitOnNumbers`?

File: source/delimiter-case.d.ts [hunk 7]
```diff
@@ -87,13 +54,9 @@ const rawCliOptions: OddlyCasedProperties<SomeOptions> = {
...
+	? string extends Value
```
Comment (som-sm) [inline source/delimiter-case.d.ts:59]: Replace the literal check with the `IsStringLiteral` type.

Right now, infinite string types (other than `string`) don't work properly, so it's best to skip them.
```ts
type T = DelimiterCase<`on${string}`, '#'>;
//   ^? type T = "on"
```

**Review by som-sm — CHANGES_REQUESTED** [review]

File: source/delimiter-case.d.ts (hunk 6 above)
Comment (filipw01) [reply source/delimiter-case.d.ts]: Added

File: source/delimiter-case.d.ts [hunk 8]
```diff
@@ -1,54 +1,20 @@
...
-			: '';
+type DelimiterCaseFromArray<
+	Words extends string[],
+	Delimiter extends string,
+	OutputString extends string = '',
+> = Words extends [
+	infer FirstWord extends string,
+	...infer RemainingWords extends string[],
+]
+	? `${Delimiter}${FirstWord}${DelimiterCaseFromArray<RemainingWords, Delimiter>}`
```
[the remark this answers is not in the document]
Comment (filipw01) [reply source/delimiter-case.d.ts:14]: Nice, catch. Changed

File: source/delimiter-case.d.ts (hunk 7 above)
Comment (filipw01) [reply source/delimiter-case.d.ts]: Added a test case and fixed it

**Comment by filipw01** [comment]
Looks good 😉

File: source/words.d.ts [hunk 9]
```diff
@@ -1,11 +1,46 @@
...
+
+	type Example2 = Words<'p2pNetwork', {splitOnNumbers: false}>;
+	//=> ["p2p", "Network"]
+	```
+	*/
+	splitOnNumbers?: boolean;
+};
+
+type DefaultOptions = {
+	splitOnNumbers: true;
```
Comment (dasa) [inline source/words.d.ts:41]: Should this default to `false`? 4.36.0 seems to be a breaking change since I have to set this to false to avoid errors using `KebabCase`.
